network/server/httpjson/handlers.go

```diff
@@ -108,23 +117,62 @@ func RegisterHandlers(mux *http.ServeMux, service interface{}, opts ServerOption
 		name := strings.ToLower(method.Name)
 		mux.HandleFunc(fmt.Sprintf("/%s", name), func(w http.ResponseWriter, r *http.Request) {
 			w.Header().Set("Content-Type", "application/json")
-			if strings.ToLower(r.Method) != "post" {
+
+			httpMethod := strings.ToUpper(r.Method)
+			if reqIn == nil && httpMethod != "GET" {
+				writeError(w, errRequestMustBeGet)
+				return
+			}
+			if reqIn != nil && httpMethod != "POST" {
 				writeError(w, errRequestMustBePost)
 				return
 			}
 
-			in := reflect.New(reqIn.Elem()).Interface()
-			defer r.Body.Close()
-			if err := json.NewDecoder(r.Body).Decode(in); err != nil {
-				writeError(w, errInvalidRequestBody)
-				return
+			headers := make(map[string]string)
+			for key, values := range r.Header {
+				if len(values) > 0 {
+					headers[key] = values[0]
+				}
 			}
 
-			svc := reflect.ValueOf(service)
+			var in interface{}
+			if reqIn != nil {
+				in = reflect.New(reqIn.Elem()).Interface()
+				defer r.Body.Close()
```

Comment:
hm..just realized this is a request body, in which case I don't think we need to close it in either case.

http://grokbase.com/t/gg/golang-nuts/138p15ng41/go-nuts-http-request-body-close